src/sparsemax.jl

```diff
@@ -0,0 +1,72 @@
+"""
+    sparsemax(x; dims = 1)
+
+[Sparsemax](https://arxiv.org/abs/1602.02068) turns input array `x`
+into sparse probability distributions that sum to 1 along the dimensions specified by `dims`.
+
+For a matrix input `x` it will by default (`dims = 1`) treat it as a batch of vectors,
+with each column independent. Keyword `dims = 2` will instead treat rows independently, and so on.
+
+# Examples
+
+```jldoctest
+julia> sparsemax([1 2 3; 2 2 2])  # dims=1
+2×3 Matrix{Float64}:
+0.0  0.5  1.0
+1.0  0.5  0.0
+
+julia> sparsemax([1 2 3; 2 2 2]; dims=2)
+2×3 Matrix{Float64}:
+0.0       0.0       1.0
+0.333333  0.333333  0.333333
+```
+"""
+sparsemax(x; dims = 1) = sparsemax!(similar(x, (float ∘ eltype)(x)), x; dims = dims)
+
+sparsemax!(x; dims = 1) = sparsemax!(x, x; dims = dims)
```

Comment:
I understand that it's probably find to have this if every dimension could be considered independent. Good to add a comment about this so it's known why it's a fine optimisation

Reply:
> What would you think of generalising this across dimensions instead checking for dims explicitly?

looking into generalizing the implementation